crates/rpc/src/server/accept.rs

```diff
@@ -75,148 +182,232 @@ where
     fn call(&mut self, request: http::Request<B>) -> Self::Future {
-        // If the ACCEPT header is not present, do not enforce the version requirement.
-        let Some(accept_header_value) = request.headers().get(ACCEPT) else {
-            return self.inner.call(request).boxed();
-        };
+        if let Err(err) = self.verifier.verify(request.headers()) {
+            let response = tonic::Status::invalid_argument(err.as_report()).into_http();
 
-        // Convert header value to str
-        let accept_str = match accept_header_value.to_str() {
-            Ok(value) => value,
-            Err(e) => {
-                debug!(target: COMPONENT, "Failed to stringify accept header value: {}", e);
-                return bad_request("Invalid accept header value".into()).boxed();
-            },
-        };
+            return futures::future::ready(Ok(response)).boxed();
+        }
 
-        // Parse the accept str
-        let accept = match AcceptHeaderValue::try_from(accept_str) {
-            Ok(value) => value,
-            Err(e) => {
-                // If the accept header value is different from expected, don't enforce the version
-                // requirement.
-                debug!(target: COMPONENT, "Failed to parse accept header value: {}", e);
-                return self.inner.call(request).boxed();
-            },
-        };
+        self.inner.call(request).boxed()
+    }
+}
 
-        // Parse the version
-        let version = match Version::parse(accept.version) {
-            Ok(version) => version,
-            Err(e) => {
-                debug!(target: COMPONENT, "Failed to parse version: {}", e);
-                return bad_request("Invalid version specified in accept header value".into())
-                    .boxed();
-            },
-        };
+#[derive(Debug, thiserror::Error)]
+enum MediaRangeParsingError {
+    #[error("media range contained no types")]
+    MissingMediaType,
+    #[error("the media type {0} is invalid")]
+    InvalidMediaType(String),
+    #[error("the parameter {0} is invalid")]
+    InvalidParameter(String),
+    #[error("the quality weight value {0} is invalid")]
+    InvalidQuality(String),
+}
 
-        // Verify the version matches configured requirements
-        if self.version_req.matches(&version) {
-            self.inner.call(request).boxed()
-        } else {
-            debug!(target: COMPONENT, "Version does not match ({}/{})", version, self.version_req);
-            let msg = format!(
-                "Client / server version mismatch. Version {} is required by the server",
-                self.version_req
-            );
-            bad_request(msg).boxed()
-        }
+#[derive(Debug)]
+struct MediaRange<'a> {
+    main: MediaType<'a>,
+    subtype: MediaType<'a>,
+    params: HashMap<CaselessKey<'a>, &'a str>,
+    quality: Quality,
+}
+
+#[derive(Debug)]
+enum MediaType<'a> {
+    Wildcard,
+    Type(&'a str),
+}
+
+#[derive(Debug, Eq, Hash)]
```

Comment:
nit: `CaselessKey("q")` and `CaselessKey("Q")` would hash to different things. Maybe you were keeping this in mind but just noting it just in case.